packages/cli/src/runner/file-command-runner.ts

```diff
@@ -2,36 +2,95 @@
+        urlToReportMap: {},
+    } as SummaryReportData;
+
+    private readonly uniqueUrls = new Set();
+
     constructor(
         @inject(AIScanner) private readonly scanner: AIScanner,
         @inject(ReportGenerator) private readonly reportGenerator: ReportGenerator,
         @inject(ReportDiskWriter) private readonly reportDiskWriter: ReportDiskWriter,
+        @inject(JsonSummaryReportGenerator) private readonly jsonSummaryReportGenerator: JsonSummaryReportGenerator,
+        @inject(ConsoleSummaryReportGenerator) private readonly consoleSummaryReportGenerator: ConsoleSummaryReportGenerator,
+        private readonly fileSystemObj: typeof fs = fs,
+        private readonly lodashObj: typeof lodash = lodash,
     ) {}
 
     public async runCommand(scanArguments: ScanArguments): Promise<void> {
-        const spinner = new Spinner('Running scanner...');
-        let axeResults: AxeScanResults;
+        const spinner = new Spinner(`Running scanner...\n`);
+        spinner.start();
+        // tslint:disable-next-line: no-any
+        const promises: any[] = [];
 
         try {
-            spinner.start();
+            const lines = this.fileSystemObj.readFileSync(scanArguments.inputFile, 'UTF-8').split(/\r?\n/);
 
-            axeResults = await this.scanner.scan(scanArguments.url);
+            lines.forEach(async (line) => {
+                // tslint:disable-next-line: no-parameter-reassignment
+                line = line.trim();
+                if (!this.lodashObj.isEmpty(line) && !this.uniqueUrls.has(line)) {
+                    this.uniqueUrls.add(line);
+                    promises.push(
+                        this.scanURL(line).then((reportContent) => {
+                            const reportName = this.reportDiskWriter.writeToDirectory(scanArguments.output, line, 'html', reportContent);
+                            this.summaryReportData.urlToReportMap[line] = reportName;
+                        }),
+                    );
+                }
+            });
         } finally {
             spinner.stop();
         }
 
-        const reportContent = this.reportGenerator.generateReport(axeResults);
+        // tslint:disable-next-line: no-floating-promises
+        Promise.all(promises).then(() => {
+            console.log(this.consoleSummaryReportGenerator.generateReport(this.summaryReportData));
+            const jsonSummryReportName = this.reportDiskWriter.writeToDirectory(
+                scanArguments.output,
+                `ViolationCountByRuleMap`,
+                'json',
+                this.jsonSummaryReportGenerator.generateReport(this.summaryReportData),
+            );
+            console.log(`ViolationCountByRuleMap summary was saved in file ${jsonSummryReportName}`);
+        });
+    }
+
+    private async scanURL(url: string): Promise<string> {
+        let axeResults: AxeScanResults;
+
+        axeResults = await this.lodashObj.cloneDeep(this.scanner).scan(url);
```

Comment:
this is not kind of object we usually clone. The scanner is setup as singleton hence we need to change to factory pattern instead

https://github.com/microsoft/accessibility-insights-service/blob/2c2391db6dd40ed17224decf35e6a2c25d776ad2/packages/cli/src/setup-cli-container.ts#L18

Reply:
the issue isn't the pattern, the issue is the way we inject our dependency, we need a new scanner for each can, but we inject the scanner in the constructor so it will be the same every for all scans, which throws exception since scan runs parallel

Reply:
you should never clone object that has functions. cloning should be done only on data object.
In this case, if you want to create multiple scanner:
1. ) you should limit the parallel scanners that are being executed. You can use https://www.npmjs.com/package/p-limit
2.) In this case you are trying to have multiple instances of scanner. I would suggest refactoring Scanner & make it stateless. (don't take page in constructor.) And also refactor Page class. (Page should not launch or close browser.)


How about something like this.:

```
const browser = webdriver.launch();

// in async with max parallel execuition of lets say 5)
urls.forEach(url => {
     await this.scanner.scan(url, browser);
})


Scanner.scan = (url: string, browser: Puppeteer.Browser) => {
    const page = this.pageFactory.create(url, browser);
    await page.create(url, browser);
    await page.scanForA11yIssues();
   ....
}



export class PageFactory {
    public create(): Page {
        return new Page(....);
    }
}
```

Reply:
We can inject factory in a constructor to use it for `AIScanner` instance creation like this for example:

https://github.com/microsoft/accessibility-insights-service/blob/2920dcab1e913e87651cad94bf9ad2f50d922314/packages/scanner/src/page.ts#L19

Reply:
no cloning anymore, updated!